### PR TITLE
Opportunist Ability

### LINKED ISSUE #2753: ABILITY_OPPORTUNIST

### PR DESCRIPTION
**Pull request by ghoulslash**
Now that mirror herb is merged, adding opportunist was relatively similar.

opportunist

I needed to add a new MOVEEND_ case and ABILITYEFFECT_ case because we needed all other ability, item, etc moveend effects to occur first. This might not be the right approach though. Would need to think if there's a simpler way to handle this. But I made the gTotemBoosts statValue setting a cumulative effect, so if two opponents have e.g. intrepid sword, the opportunist mon will get a sharp attack boost (see below) - This needs more research as far as official mechanics. Would opportunist trigger after each individual intrepid sword?

opportunist_2

opportunist3

Closes #2753

### CODE REVIEW

**Review by AgustinGDLV — COMMENTED** [review]
Looks great on my local branch! I'm a little put off by `gTotemBoosts` staying the name for Mirror Herb and Opportunist cases. I think it's fine that we're reusing it, and I understand that this was already approved for Mirror Herb, so I'll let you decide if it should be changed to something a bit more accurate. I can only think of something like `gQueuedStatBoosts`, but that might be a bit misleading, too. 

Illegal Ability Tests

**Comment by AsparagusEduardo** [comment]
I agree, `gQueuedStatBoosts` would be a better name for the field now that it's used for other stuff than Totem boosts.

**Comment by ghoulslash** [comment]
> I agree, `gQueuedStatBoosts` would be a better name for the field now that it's used for other stuff than Totem boosts.

Done

**Comment by AsparagusEduardo** [comment]
Can you merge with upcoming so that it passed the checks thanks to #3045?

**Comment by AsparagusEduardo** [comment]
It's failing due to a different reason 👀

**Comment by ghoulslash** [comment]
branch updated and test fixed.ready for re-review

**Review by DizzyEggg — COMMENTED** [review]
Just a comment for now. 
Would you mind creating a test for Opportunist in a double battle where partner has Intimidate, but one of the foes has Contrary, so its Attack rises. This seems like an edge case that could explode, but if it works, then I'd feel pretty comfortable merging it.

**Comment by ghoulslash** [comment]
> Just a comment for now. 
> Would you mind creating a test for Opportunist in a double battle where partner has Intimidate, but one of the foes has Contrary, so its Attack rises. This seems like an edge case that could explode, but if it works, then I'd feel pretty comfortable merging it.

As per issue #3356 that given test will fail but not because of opportunist. Do we want to make the test now or add it to whichever PR fixes the bug?

**Comment by mrgriffin** [comment]
> As per issue #3356 that given test will fail but not because of opportunist. Do we want to make the test now or add it to whichever PR fixes the bug?

I would add it now, but marked with [`KNOWN_FAILING`](<https://github.com/rh-hideout/pokeemerald-expansion/blob/master/include/test/battle.h#L202>).

File: test/battle/ability/opportunist.c [hunk 1]
```diff
@@ -0,0 +1,41 @@
+#include "global.h"
+#include "test/battle.h"
+
+ASSUMPTIONS
+{
+    ASSUME(gSpeciesInfo[SPECIES_WOBBUFFET].abilities[2] == ABILITY_OPPORTUNIST);
```
Comment (AsparagusEduardo) [inline test/battle/ability/opportunist.c:6]: Let's not add false assumptions ^^;

File: test/battle/ability/opportunist.c [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        EXPECT_MUL_EQ(results[0].damage, Q_4_12(2.0), results[1].damage);
+        // stat boosts should be the same
+        EXPECT_EQ(player->statStages[STAT_ATK], opponent->statStages[STAT_ATK]);
+        EXPECT_EQ(player->statStages[STAT_SPATK], opponent->statStages[STAT_SPATK]);
+        EXPECT_EQ(player->statStages[STAT_SPEED], opponent->statStages[STAT_SPEED]);
+        // opportunist should not copy stat drops from shell smash
+        EXPECT_LT(player->statStages[STAT_DEF], opponent->statStages[STAT_DEF]);
+        EXPECT_LT(player->statStages[STAT_SPDEF], opponent->statStages[STAT_SPDEF]);
+    }
+}
```
Comment (AsparagusEduardo) [inline test/battle/ability/opportunist.c:40]:
```suggestion
}

TO_DO_BATTLE_TEST("Opportunist doesn't copy ally stat increases");
TO_DO_BATTLE_TEST("Opportunist doesn't copy foe stat increases gained via Opportunist");
TO_DO_BATTLE_TEST("Opportunist copies foe stat increased gained via Swagger and Flatter");
```

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

**Comment by ghoulslash** [comment]
Added contrary + intimidate + opportunist test. ready to review

File: test/battle/ability/opportunist.c [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            MESSAGE("Wobbuffet copied its opponent's stat changes!");
+            ANIMATION(ANIM_TYPE_GENERAL, B_ANIM_STATS_CHANGE, playerRight);
+            MESSAGE("Wobbuffet's Attack sharply rose!");
+        }
+        
+        HP_BAR(playerLeft, captureDamage: &results[i].damageLeft);
+        HP_BAR(playerRight, captureDamage: &results[i].damageRight);
+    } THEN {
+        EXPECT_EQ(opponentLeft->statStages[STAT_ATK],  (abilityLeft == ABILITY_CONTRARY)  ? DEFAULT_STAT_STAGE+1 : DEFAULT_STAT_STAGE-1);
+        EXPECT_EQ(opponentRight->statStages[STAT_ATK], (abilityRight == ABILITY_CONTRARY) ? DEFAULT_STAT_STAGE+1 : DEFAULT_STAT_STAGE-1);
```
Comment (AsparagusEduardo) [inline test/battle/ability/opportunist.c:90]:
```suggestion
        EXPECT_EQ(opponentLeft->statStages[STAT_ATK],  (abilityLeft == ABILITY_CONTRARY)  ? DEFAULT_STAT_STAGE + 1 : DEFAULT_STAT_STAGE - 1);
        EXPECT_EQ(opponentRight->statStages[STAT_ATK], (abilityRight == ABILITY_CONTRARY) ? DEFAULT_STAT_STAGE + 1 : DEFAULT_STAT_STAGE - 1);
```

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

**Review by AsparagusEduardo — APPROVED** [review]
